### ISSUE TITLE
Reach out to Travis to get option that includes

### ISSUE THREAD

**Issue by mhdawson** [nodejs/package-maintenance#205]
LTS releases + current.

**Comment by ljharb** [comment]
This is already doable by adding matrix lines in `.travis.yml` for each LTS signifier that nvm supports, as well as `node`. I'm not sure how it would work to have something in the yml that created more than one item.

**Comment by dominykas** [comment]
Yes, it is possible to do this and we should evangelize that, however I think this issue is to change the default, i.e. the behavior when you have _no_ versions defined. At the moment it will only run in the latest stable - I think we should advocate that it should run in the latest stable + latest LTS and that is _without_ adding any lines other than `language: node_js` in the `.travis.yml`.

I also think that `lts/*` was _intended_ to support multiple LTS versions, but maybe they never got around to implementing that.

So, to summarize, what I'd really love to see travis do is support the following:

- `node` latest stable Node.js release (same as they currently)
- `lts/*` all active LTS Node.js releases (currently:  latest LTS Node.js release)
- default: `current` (or `active` or smth like that) equivalent of having both `node` and `lts/*` in the version list (currently: default is `node`).

Docs: https://docs.travis-ci.com/user/languages/javascript-with-nodejs/#specifying-nodejs-versions

**Comment by ljharb** [comment]
I’m the maintainer of nvm who invented “lts/*” - in no way was it ever intended to support multiple versions, it’s a single star and behaves the same way a single star does on the command line, selecting the first matched version only.

**Comment by dominykas** [comment]
Cool, thanks, good to know! Did not realize this comes from `nvm` and assumed it's a glob.

**Comment by ljharb** [comment]
That would still indicate a single version; nvm (travis’ sole means of deciding node version) does not have any concept of selecting multiple versions, only listing them.

I can certainly ask Travis to change the default matrix to have multiple jobs, but i anticipate pushback about the complexity for users of getting, say, 3 jobs by default, adding 1, and ending up with 1 - if that comes up, how should i respond?

**Comment by dominykas** [comment]
Ideally, library maintainers, esp. for popular libs, _should_ be testing at least in all active LTS plus stable, but if that's not feasible (there's an obvious cost associated with 3x the job runs), then the current default of `stable` is a good one, I would say.

Regardless of that, adding a way to define "all active LTS" in a single line of `.travis.yml`, which does not need to be maintained, would be an awesome feature.

And even without that feature, possibly Travis could update their docs to highlight the current behavior (that people might not be testing in LTS by default) and include a recommendation and an example of a manual configuration. I'm happy to try and work on a proposal for the docs (not sure if they have them in the open so people can PR).

**Comment by mhdawson** [comment]
I think the 

> which does not need to be maintained,

part would really be helpful.  @ljharb if a single entry mapping to multiple entries does not fit the model any other ideas?  I wonder about something like LTS-latest, LTS-1, LTS-2 or some other way to try and catch all of the active LTS versions which would not have to be updated for each new release.

**Comment by ljharb** [comment]
hmm, probably the easiest way would be for me to add a similar feature to nvm; then travis can get it by default.

**Comment by dominykas** [comment]
@ljharb any chance there's an update on this? I'd be keen to hear any and all opinions from Travis, before I start drafting #204... Can I be of any assistance on the nvm side of things (a stretch, given I'm no bash guru, but still)?

**Comment by dominykas** [comment]
By the way, renovate does have some syntax on this: https://renovatebot.com/node/

**Comment by ljharb** [comment]
@dominykas the more i think about it, the more i don't think it's really compatible with travis without being built into nvm. (the renovate link 404s for me)

**Comment by dominykas** [comment]
Looks like a bug in renovate docs routing or smth, here's a link that works: https://renovatebot.com/docs/node/

**Comment by ghinks** [comment]
I left a write up at [206](https://github.com/nodejs/package-maintenance/issues/206) and the **renovate** bot does appear to meet our needs.

**Comment by dominykas** [comment]
Travis team page: https://about.travis-ci.com/team/

Not sure whether it's appropriate to tag people that are also on GitHub directly into this issue :)

**Comment by dominykas** [comment]
Just got a word from Travis that they're changing the default to be the LTS: https://github.com/travis-ci/travis-build/pull/1747 (this ofc means that most packages won't be testing with the next LTS, so there's still work to be done and evangelized).

**Comment by mhdawson** [comment]
From my reading I think lts/* is only the latest LTS and we were thinking a default for all current LTS releases plus the current release would be what we'd want to be suggesting.

**Comment by ljharb** [comment]
I believe that would be a massive architectural change for travis to have a default matrix instead of a single default job.

**Comment by mhdawson** [comment]
For for the confusion, should have re-read the issue to refresh my memory.  I can see that the default being to test all LTS versions and current could be impractical, but a way that it could be selected easily without having to update it every time a new LTS release comes out would still be useful.  Sounds like the easiest way would still be an option in nvm?

**Comment by ljharb** [comment]
`lts/*` always live-updates to whatever the latest LTS is based on nodejs.org/dist/index.tab

**Comment by mhdawson** [comment]
Right so that covers the latest, but we'd want something similar to the previous LTS version as well as there is usually 2 active LTS releases, if not 3.

**Comment by ljharb** [comment]
nvm could certainly add `lts/*-2` or something similar, but that still wouldn't make it practical to have travis default to testing more than one node version.

**Comment by mhdawson** [comment]
But it would make it easier for a maintainer to setup their testing to test all LTS versions and Current once and then not have to update it for each LTS release right?

**Comment by ljharb** [comment]
If they hardcoded the assumption that there's 2 active LTS lines, then yes.

**Comment by ljharb** [comment]
Filed https://github.com/nodejs/Release/issues/464 as well, which would be helpful/needed for this.

**Comment by dominykas** [comment]
> I believe that would be a massive architectural change for travis to have a default matrix instead of a single default job.

Is that and assumption we should be making though? The list of versions to run is just a key in the `.travis.yaml`. Presumably when the key is missing - some default value is returned from somewhere. That somewhere is hopefully the only place that needs to change? Well, in an ideal world. We shouldn't make this assumption either, of course.

I've asked the support to check if there's a chance they could join us for more discussions, to clear up these assumptions and to see what's possible.

> nvm could certainly add lts/*-2 or something similar

This is https://github.com/nvm-sh/nvm/issues/1998, right?

**Comment by ljharb** [comment]
@dominykas when you add your first line to travis.yml, your matrix shouldn't go from 4 jobs to 1.

Yep, that's the issue to track!

**Comment by dominykas** [comment]
> when you add your first line to travis.yml, your matrix shouldn't go from 4 jobs to 1.

That's a question that's probably best answered by some usability testing. But with a smart default you wouldn't need to add such a line, i.e. the line would be left for "advanced" users and can also be tackled by good documentation and UI.

There's also the case to be made for migrating away from`node_js` version list to a `node_js_version_policy` type thing (which would then expand to the `node_js`, taking precedence over the defaults - once again - assumptions - but not entirely implausible).

The `node_js_version_policy` could then take a single string, which we could standardize to what renovate already supports: [`all`, `lts`, `active`, `lts_active`, `lts_latest`, `current`](https://renovatebot.com/docs/node/) (or at least use that as a baseline, because "latest" vs "current" is rather ambiguous). The same policy string should be used in `support.json`, btw. Not sure where's the best place to start an issue/discussion/doc on this? Here?

**Comment by nschonni** [comment]
I would think maybe over here https://travis-ci.community/c/languages/nodejs

**Comment by wesleytodd** [comment]
With the new announcement of Gihub CI, we might also want to reach out to them to get the same "default" behavior for node environments we are working on for Travis.  I am not exactly sure how it will work, but it looks like there will be some built in steps you can leverage (from the landing page `- uses: actions/setup-node@master`).  If we can have some say in how these work it would set a good precedent for all the users who will probably jump on board with this.

**Comment by wesleytodd** [comment]
An issue to track the alias' they will add.  Probably a good place to get our opinions heard: https://github.com/actions/setup-node/issues/26

**Comment by chrispat** [comment]
@wesleytodd I have not followed this thread in detail but would be curious to understand what recommendations the node community has.  The starter workflow for node.js is  in the [actions/starter-workflows](https://github.com/actions/starter-workflows/blob/master/ci/node.js.yml) repo. We would welcome any pull request for suggestions to improve it.

**Comment by wesleytodd** [comment]
Hi @chrispat, the main goal is to improve the story around package authors testing their packages in all of the current node versions.  It is very typical for authors to set travis to test the important versions at the time of publishing, but that is quickly out of date.  The PR I made above is just to get users started, but what we really need is an alias.  It sounds like, from the people who commented on those issues I linked above, this is on the roadmap.  So I don't know if there is more to do at this time from the GH side.

**Comment by dominykas** [comment]
@chrispat I will be writing a blog post on what people should be testing in, but in short, I think the recommendation is to _at least_ test in the latest version of the current LTS release lines (currently: 8.16.0 and 10.16.2), but ideally to test in all the actively maintained release lines (currently: 8.16.0, 10.16.2, 12.8.0; in Apr/2019: latest versions of 6.x, 8.x, 10.x, 11.x).

Ideally these lists would be available via some alias, so that people do not have to keep that updated. There is a discussion in #236 on what keywords to use for such aliases.

**Comment by mhdawson** [comment]
Waiting on

* Issue in nvm to provide option for oldest LTS release
* Issue in Travis to adopt the above
* (blog post already complete)

**Comment by ljharb** [comment]
This can now be achieved trivially by using my [shared configs](https://github.com/ljharb/travis-ci): https://twitter.com/ljharb/status/1194728750698487808

iow, anyone who uses my shared travis configs will get new LTS versions automatically, since i add new node versions to that repo pretty rapidly (and eventually could automate it with github actions).

**Comment by dominykas** [comment]
👍 This is great!

What about deprecating old versions?

**Comment by ljharb** [comment]
Happy to add any additional composable "presets" to that shared repo as folks need it.

**Comment by mhdawson** [comment]
Very nice !

**Comment by ljharb** [comment]
as an example, node v13.2 just was released, so now every project using the "minors" presets will automatically test on it also (as opposed to instead of 13.1), on their next travis run.

**Comment by dominykas** [comment]
1. Should we have this under `@pkgjs`? I know @ljharb is not some rando on the internet, but if we were to make an official recommendation - should that be in an official repo?
2. It does a bit more than just include the relevant versions (e.g. `npm run tests-only` support)? Do we want to be that opinionated? I think I'd favor something that has just the basics.
3. It would probably make sense to provide includes named based on the keyword convention we have for package support?

**Comment by mhdawson** [comment]
I'd be +1 for having it under #pkgjs

**Comment by ljharb** [comment]
I'd be happy to fork it into an official org, and then keep both remotes updated at the same time.

**Comment by ljharb** [comment]
As for the `tests-only` thing, that's kind of necessary if we want to separate linting, from auditing, from tests - since `npm test` is supposed to always run 100% of your tests, I've chosen "pretest", "posttest", and "tests-only" as the convention. There always has to be *some* convention since there's no way to do `npm test --ignore-scripts`, unfortunately.

**Comment by dominykas** [comment]
If I understand it correctly, Travis can import just the `nodejs` bit - it doesn't _need_ to have the `script`, `env`, etc? Similarly, while I personally also prefer to have `nvm install-latest-npm` in my scripts, I would not say that this is the 100% default for everyone?

**Comment by ljharb** [comment]
That's a fair point; I can separate out the matrices, and the "latest npm", to separate imports, to allow people to pick and choose those and define their own script/pretest/posttest settings.

Separately tho, I'm not sure what use case anyone has in CI for not testing with the latest npm :-)

**Comment by dominykas** [comment]
> Separately tho, I'm not sure what use case anyone has in CI for not testing with the latest npm :-)

I can come up with several edge cases, alright... Not sure if they're relevant for people publishing to public npm, though. One example is doing stuff with docker - the base images have a fixed npm which is distributed with the node release. Then there's occasional edge cases for more complex situations with peer deps and/or shrinkwraps, which are about to get breaking changes and ever so often have regressions. Then there's the logic of using the same npm your users are likely to use (i.e. default one bundled with node - although I'd love to see some stats on this!)

Anyways, like I said - I'm with you on this one ;) just not sure if we want to make it the default.

Note to self: we should update best practices to include this, if they don't already do that.

**Comment by ljharb** [comment]
I’ll comment here once I’ve separated out the unopinionated parts.

**Comment by dominykas** [comment]
We now have https://github.com/nodejs/ci-config-travis - if there's no objections, I'll close this issue in a week or so?
